pkg/controller/migmigration/task.go

```diff
@@ -27,6 +29,95 @@ const (
 	Completed               = "Completed"
 )
 
+var PhaseOrder = map[string]int{
+	Created:                 00, // 0-499 normal
+	Started:                 1,
+	WaitOnResticRestart:     10,
+	ResticRestartCompleted:  11,
+	BackupStarted:           20,
+	BackupCompleted:         21,
+	WaitOnBackupReplication: 30,
+	BackupReplicated:        31,
+	RestoreStarted:          40,
+	RestoreCompleted:        41,
+	BackupFailed:            501, // 500-999 errors
+	RestoreFailed:           510,
+	Completed:               1000, // Succeeded
+}
+
+// Phase
+type Phase struct {
+	Name string
+}
+
+// Validate the phase.
+func (p Phase) Validate() error {
+	_, found := PhaseOrder[p.Name]
+	if !found {
+		m := fmt.Sprintf("phase not valid: %s", p.Name)
+		return errors.New(m)
+	}
+
+	return nil
+}
+
+// Set the phase.
+// Prevents rewind.
+func (p *Phase) Set(name string) {
+	if p.Before(name) {
+		p.Name = name
+	}
+}
+
+// Phase is equal.
+func (p Phase) Equals(other string) bool {
+	return p.Name == other
+}
+
+// Phase is after `other`.
+func (p Phase) After(other string) bool {
+	nA, found := PhaseOrder[p.Name]
+	if !found {
+		nA = 0
+	}
+	nB, found := PhaseOrder[other]
+	if !found {
+		nB = 0
+	}
+	return nA > nB
+}
+
+// Phase is before `other`.
+func (p Phase) Before(other string) bool {
+	nA, found := PhaseOrder[p.Name]
+	if !found {
+		nA = 0
+	}
+	nB, found := PhaseOrder[other]
+	if !found {
+		nB = 0
+	}
+	return nA < nB
+}
+
+// Phase is final.
+func (p Phase) Final() bool {
+	n, found := PhaseOrder[p.Name]
+	if !found {
+		n = 0
+	}
+	return n > 500
+}
+
+// Phase is final.
+func (p Phase) Failed() bool {
+	n, found := PhaseOrder[p.Name]
+	if !found {
+		n = 0
+	}
+	return n >= 500 && n < 1000
```

Comment:
In one place the check is `>500` and in another it's `>=500` Probably should be one or the other -- is 500 considered a failed condition or do failed conditions start at 501?

Reply:
Good catch.  500 is the first error code and the comparison should be `>= 500`.